ts_src/payments/p2tr.ts

```diff
@@ -9,275 +9,285 @@ import {
-      return bech32m.encode(network.bech32, words);
-    });
-
-    lazy.prop(o, 'hash', () => {
-      if (a.hash) return a.hash;
-      if (a.scriptsTree) return toHashTree(a.scriptsTree).hash;
-      const w = _witness();
-      if (w && w.length > 1) {
-        const controlBlock = w[w.length - 1];
-        const leafVersion = controlBlock[0] & 0b11111110;
-        const script = w[w.length - 2];
-        const leafHash = tapLeafHash(script, leafVersion);
-        return rootHashFromPath(controlBlock, leafHash);
-      }
-      return null;
-    });
-    lazy.prop(o, 'output', () => {
-      if (!o.pubkey) return;
-      return bscript.compile([OPS.OP_1, o.pubkey]);
-    });
-    lazy.prop(o, 'scriptLeaf', () => {
-      if (a.scriptLeaf) return a.scriptLeaf;
-    });
-    lazy.prop(o, 'pubkey', () => {
-      if (a.pubkey) return a.pubkey;
-      if (a.output) return a.output.slice(2);
-      if (a.address) return _address().data;
-      if (o.internalPubkey) {
-        const tweakedKey = tweakKey(o.internalPubkey, o.hash);
-        if (tweakedKey) return tweakedKey.x;
-      }
-    });
-    lazy.prop(o, 'internalPubkey', () => {
-      if (a.internalPubkey) return a.internalPubkey;
-      const witness = _witness();
-      if (witness && witness.length > 1)
-        return witness[witness.length - 1].slice(1, 33);
-    });
-    lazy.prop(o, 'signature', () => {
-      if (!a.witness || a.witness.length !== 1) return;
-      return a.witness[0];
-    });
-    lazy.prop(o, 'input', () => {
-      // todo
-    });
-    lazy.prop(o, 'witness', () => {
-      if (a.witness) return a.witness;
-      if (a.scriptsTree && a.scriptLeaf && a.internalPubkey) {
-        // todo: optimize/cache
-        const hashTree = toHashTree(a.scriptsTree);
-        const leafHash = tapLeafHash(a.scriptLeaf.output, a.scriptLeaf.version);
-        const path = findScriptPath(hashTree, leafHash);
-        const outputKey = tweakKey(a.internalPubkey, hashTree.hash);
-        if (!outputKey) return;
-        const version = a.scriptLeaf.version || 0xc0;
-        const controlBock = NBuffer.concat(
-          [NBuffer.from([version | outputKey.parity]), a.internalPubkey].concat(
-            path.reverse(),
-          ),
-        );
-        return [a.scriptLeaf.output, controlBock];
-      }
-      if (a.signature) return [a.signature];
-    });
-
-    // extended validation
-    if (opts.validate) {
-      let pubkey: Buffer = NBuffer.from([]);
-      if (a.address) {
-        if (network && network.bech32 !== _address().prefix)
-          throw new TypeError('Invalid prefix or Network mismatch');
-        if (_address().version !== TAPROOT_VERSION)
-          throw new TypeError('Invalid address version');
-        if (_address().data.length !== 32)
-          throw new TypeError('Invalid address data');
-        pubkey = _address().data;
-      }
+export function p2tr(
+  a: Payment,
+  opts?: PaymentOpts,
+  eccLib?: TinySecp256k1Interface,
```

Comment:
I was thinking it should be a member of the first argument `a`

Reply:
- `a` and the output `o` are related data structures (`o` is `a` after the computations and validations have been applied)
  - if `eccLib` is not expected to be present on `o` then it shouldn't be on `a` either
- the `PaymentOpts` looks be a better place for `eccLib` (`eccLib` is actually an option for the payment)

Reply:
good point. PaymentOpts sounds better

Reply:
> good point. PaymentOpts sounds better
Done!